lib/auth/join_test.go

```diff
@@ -367,3 +367,81 @@ func TestRegister_Bot(t *testing.T) {
 		})
 	}
 }
+
+// TestRegister_Bot_Expiry checks that bot certificate expiry can be set, and
+// does not exceed the limit.
+func TestRegister_Bot_Expiry(t *testing.T) {
+	t.Parallel()
+	ctx := context.Background()
+
+	srv := newTestTLSServer(t)
+	privateKey, publicKey, err := testauthority.New().GenerateKeyPair()
+	require.NoError(t, err)
+	sshPrivateKey, err := ssh.ParseRawPrivateKey(privateKey)
+	require.NoError(t, err)
+	tlsPublicKey, err := tlsca.MarshalPublicKeyFromPrivateKeyPEM(sshPrivateKey)
+	require.NoError(t, err)
+
+	validExpires := srv.Clock().Now().Add(time.Hour * 6)
+	tooGreatExpires := srv.Clock().Now().Add(time.Hour * 24 * 365)
+	tests := []struct {
+		name           string
+		requestExpires *time.Time
+		expectTTL      time.Duration
+	}{
+		{
+			name:           "unspecified defaults",
+			requestExpires: nil,
+			expectTTL:      defaults.DefaultRenewableCertTTL,
+		},
+		{
+			name:           "valid value specified",
+			requestExpires: &validExpires,
+			expectTTL:      time.Hour * 6,
+		},
+		{
+			name:           "value exceeding limit specified",
+			requestExpires: &tooGreatExpires,
+			// MaxSessionTTL set in createBotRole is 12 hours, so this cap will
+			// apply instead of the defaults.MaxRenewableCertTTL specified
+			// in generateInitialBotCerts.
+			expectTTL: 12 * time.Hour,
```

Comment:
This is a little unexpected. Looking at the implementation of `generateInitialBotCerts` you'd expect this to be 24 hours as this value is enforced as a maximum, but the bot role specifies a `MaxSessionTTL` of 12 hours, so this overrides the 24 hour value that is enforced.

Reply:
@timothyb89 - I can see a TODO in `createBotRole` - is this something we should reconsider now ?